server/src/main/java/org/elasticsearch/index/mapper/IpFieldMapper.java

```diff
@@ -466,8 +474,20 @@ protected void parseCreateField(DocumentParserContext context) throws IOExceptio
     }
 
     private void indexValue(DocumentParserContext context, InetAddress address) {
+        // Check that a dimension field is single-valued and not an array
+        if (dimension && context.doc().getByKey(fieldType().name()) != null) {
+            throw new IllegalArgumentException("Dimension field [" + fieldType().name() + "] cannot be a multi-valued field.");
+        }
         if (indexed) {
-            context.doc().add(new InetAddressPoint(fieldType().name(), address));
+            Field field = new InetAddressPoint(fieldType().name(), address);
+
+            if (dimension) {
+                // Add dimension field with key so that we ensure it is single-valued.
+                // Dimension fields are always indexed.
+                context.doc().addWithKey(fieldType().name(), field);
+            } else {
+                context.doc().add(field);
+            }
```

Comment:
I see why you did it this way. I wish there were something cleaner - the `addWithKey` thing does seem to be something we do in other places though.

I wonder if we need the `getByKey` at all - maybe we could make `addWithKey` return the old field and we could check and throw an exception we like? Or something like that. I dunno, feels wasteful to check up front when we already have the check in `addWithKey`.

Reply:
I could totally omit the `getByKey()` call. `addWithKey()` checks if there is an entry already and will throw a `throw new IllegalStateException("Only one field can be stored per key");` 

I used `getByKey()` because I wanted to produce a better error message

Reply:
It looks like we do the `if (context.doc().getByKey(fieldType().name()) != null) {` dance all over the place. Would you be ok refactoring them all in a follow up? It feels like a good thing to clean up to me but not a good thing to mix into this change.

Reply:
If we're going to keep the `getByKey` thing in this change, could you move it to right above `addWithKey` line? I feel like putting them together would make it a little more readable.